vm/src/obj/objdict.rs

```diff
@@ -174,231 +173,146 @@ fn dict_new(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
     Ok(dict)
 }
 
-fn dict_len(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
-    arg_check!(vm, args, required = [(dict_obj, Some(vm.ctx.dict_type()))]);
-    let elements = get_elements(dict_obj);
-    Ok(vm.ctx.new_int(elements.len()))
-}
+impl PyDictRef {
+    fn len(self, _vm: &mut VirtualMachine) -> usize {
+        self.entries.borrow().len()
+    }
 
-fn dict_repr(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
-    arg_check!(vm, args, required = [(dict_obj, Some(vm.ctx.dict_type()))]);
-
-    let s = if let Some(_guard) = ReprGuard::enter(dict_obj) {
-        let elements = get_key_value_pairs(dict_obj);
-        let mut str_parts = vec![];
-        for (key, value) in elements {
-            let key_repr = vm.to_repr(&key)?;
-            let value_repr = vm.to_repr(&value)?;
-            let key_str = objstr::get_value(&key_repr);
-            let value_str = objstr::get_value(&value_repr);
-            str_parts.push(format!("{}: {}", key_str, value_str));
-        }
+    fn repr(self, vm: &mut VirtualMachine) -> PyResult {
+        let s = if let Some(_guard) = ReprGuard::enter(self.as_object()) {
+            let elements = get_key_value_pairs(self.as_object());
+            let mut str_parts = vec![];
+            for (key, value) in elements {
+                let key_repr = vm.to_repr(&key)?;
+                let value_repr = vm.to_repr(&value)?;
+                let key_str = objstr::get_value(&key_repr);
```

Comment:
(we probably should change `to_repr()` to return `PyResult<PyStringRef>` - `get_value` calls are a bit ugly and they panic when `__repr__` returns a non-string.)

Reply:
Have changed both ```vm.to_str``` and ```vm.to_repr```.